keyedcache/api.go

```diff
@@ -0,0 +1,153 @@
+//     of the object)
+//   o Free() -- called when the cache is no longer referencing the object;
+//     it cannot call back into this instance of the cache (I'm not sure that
+//     this method makes sense in Go).
+//   o SizeOf() -- return memory used by the object (to track memory consumption);
+//     all objects are assumed to be the same size.
+//
+// Note: perhaps the Read() and Write() operations should be asynchronous.
+//
+type Object interface {
+	Read(cntxt *ReqContext, key *Key) (err error)
+	Write(cntxt *ReqContext, key *Key) (err error)
+	Inval(cntxt *ReqContext, key *Key)
+	Free(cntxt *ReqContext)
+	SizeOf() uint64
+}
+
+// CachedObject holds the information that a keyedcache uses to manage an object
+// stored in the cache.  An object managed by the cache must include an embedded
+// pointer to this interface, which it must initializ with a pointer passed to
+// the NewObject function supplied by the client of the cache (so when an object
+// is created by NewObject() this interface is initialized).
+//
+// CachedObject supports the following methods. These may be invoked by a client
+// that has a hold on the object.  Successful lookup of an object in the cache
+// returns a pointer to a held object.
+//
+//   o Hold() -- get an additional hold on the object
+//   o Release() -- release a hold on the object; once all holds on an object
+//     are released the object may be invalidated
+//   o MarkDirty() -- mark a held object dirty; dirty objects are cleaned
+//     (written) by the cache before invalidation unless a forced invalidation
+//     is done
+//   o IsDirty() -- returns true if the object is dirty
+//   o Flush() -- cause a dirty object to be flushed via a call to Write();
+//     panics if the object is not dirty
+//
+// If Flush() fails because the corresponding Write() fails then the object is
+// marked dirty.
+//
+type CachedObject interface {
+	Hold(cntxt *ReqContext)
+	Release(cntxt *ReqContext)
+	MarkDirty(cntxt *ReqContext)
+	IsDirty(cntxt *ReqContext) bool
+	Flush(cntxt *ReqContext) (err error)
+}
+
+// Cache caches objects identified by keys, upto a specified maximum amount of
+// memory.  Objects in the cache are accessed by calling Lookup() with the
+// desired key and Lookup() returns a pointer to an object with a hold on it.
+// When the caller is done with the object it must release the hold.
+//
+// If the object is not currently cached, Cache will create a new object
+// corresponding to the key or reuse an existing one.  It is initialized by a
+// call to Object.Read() before return.  Cache locks the object duing the call
+// to Object.Read() so only one object with a given key is present in the cache
+// and no thread can access it until it is initialized.
+//
+// There is no provision for otherwise locking the object.  In particular, an
+// object can be marked dirty or flushed while multiple threads have a hold on
+// the object.
+//
+// A Cache provides the following methods:
```

Comment:
So no "iterate through"?
I could imagine a desire to "flush everything" wanting an iterator.
Indeed, it might be interesting to iterate just over the dirty entries or just over the clean entries...